ludwig/utils/backward_compatibility.py

```diff
@@ -68,25 +109,40 @@ def _traverse_dicts(config: Any, f: Callable[[Dict], None]):
             _traverse_dicts(v, f)
 
 
-def _upgrade_use_bias(config):
-    if BIAS in config:
-        warnings.warn('Parameter "bias" renamed to "use_bias" and will be removed in v0.6', DeprecationWarning)
-        config[USE_BIAS] = config[BIAS]
-        del config[BIAS]
-    if CONV_BIAS in config:
-        warnings.warn(
-            'Parameter "conv_bias" renamed to "conv_use_bias" and will be removed in v0.6', DeprecationWarning
-        )
-        config[CONV_USE_BIAS] = config[CONV_BIAS]
-        del config[CONV_BIAS]
-    if DEFAULT_BIAS in config:
-        warnings.warn(
-            'Parameter "default_bias" renamed to "default_use_bias" and will be removed in v0.6', DeprecationWarning
-        )
-        config[DEFAULT_USE_BIAS] = config[DEFAULT_BIAS]
-        del config[DEFAULT_BIAS]
+@register_config_transformation("0.5")
```

Comment:
What do you think about a slightly modified design where:
- config transformations are registered with the `warning` message, e.g.

    ```python
    @register_config_transformation("0.5", deprecation_warning='Config section "training" renamed to "trainer" and will be removed in v0.6')
    ```

- config transformations return a tuple `(config, bool)`, with the `bool` indicating whether the transformation was actually applied.
- `VersionTransformation.transform_config()` logs all of the deprecation warnings in one place, anytime `t.transform_config()` returns `(config, True)`.

A couple of benefits of this design is that it would encourage better encapsulated transformations, and centralize deprecation logging. WDYT?

Reply:
I like the idea of requiring a deprecation message, thats a great idea!  However, this would force us to up to break up `_upgrade_feature`, `_upgrade_hyperopt`, `_upgrade_preprocessing_defaults`, and any other function which potentially makes multiple edits to the config.

It also would make it harder to use format strings in logs messages, i.e.
```
    f'executor type "{executor_type}" not supported, converted to "ray" will be flagged as error '
    "in v0.6",
```

We could enforce a convention "one config edit per function", but I'd rather have a looser guideline like "one conceptual config change per function" and its up to the author and reviewer to make that decision.